src/parser/jobs/ast/modules/Lightspeed.tsx

```diff
@@ -0,0 +1,234 @@
+	deathTruncated: boolean
+}
+
+// in this module we only want to track Lightspeed windows opened by
+// the character selected for analysis. windows that clip into
+// AST Lightspeed will be marked.
+// Used DNC Technicalities as basis for this module. Rewritten from previous module for consistency purposes
+export default class Lightspeed extends Analyser {
+	static override handle = 'Lightspeed'
+	static override title = t('ast.Lightspeed.title')`Lightspeed`
+	static override displayOrder = DISPLAY_ORDER.LIGHTSPEED
+
+	@dependency private timeline!: Timeline
+	@dependency private data!: Data
+
+	private history: LIGHTSPEED_Window[] = []
+	private currentWindow: LIGHTSPEED_Window | undefined = undefined
+	private castHook?: EventHook<Events['action']>
+
+	override initialise() {
+		const lightspeedFilter = filter<Event>().status(this.data.statuses.LIGHTSPEED.id)
+
+		this.addEventHook(lightspeedFilter.type('statusApply')
+			.target(this.parser.actor.id), this.tryOpenWindow)
+		this.addEventHook(lightspeedFilter.type('statusRemove')
+			.target(this.parser.actor.id), this.tryCloseWindow)
+
+		//this.addEventHook(filter<Event>().source(this.parser.actor.id).type('action'), this.onCast)
+	}
+
+	private tryOpenWindow(event: Events['statusApply']) {
+		if (this.currentWindow === undefined) {
+			this.currentWindow = {
+				start: event.timestamp,
+				rotation: [],
+				gcdCount: 0,
+				mpSavings: 0,
+
+				buffsRemoved: [],
+				deathTruncated: false,
+
+			}
+			this.castHook = this.addEventHook(
+				filter<Event>()
+					.source(this.parser.actor.id)
+					.type('action'),
+				this.onCast,
+			)
+		}
+
+		// If that happens, re-open the last window and keep tracking
+		if (this.currentWindow.end != null && this.currentWindow.end === event.timestamp) {
+			this.currentWindow.end = undefined
+		}
+		return this.currentWindow
+	}
+
+	private tryCloseWindow(event: Events['statusRemove']) {
+
+		if (this.currentWindow == null) {
+			return
+		}
+
+		// Cache whether we've seen a buff removal event for this status, just in case they happen at exactly the same timestamp
+		this.currentWindow.buffsRemoved.push(event.status)
+
+		// Make sure all applicable statuses have fallen off before the window closes
+		if (this.currentWindow.buffsRemoved.includes(this.data.statuses.LIGHTSPEED.id)) {
+			this.currentWindow.end = event.timestamp
+			this.history.push(this.currentWindow)
+			this.currentWindow = undefined
+			if (this.castHook != null) {
+				this.removeEventHook(this.castHook)
+				this.castHook = undefined
+			}
+		}
+	}
+
+	private onCast(event: Events['action']) {
+
+		// If we don't have a window, bail
+		if (this.currentWindow == null) {
+			return
+		}
+
+		const action = this.data.getAction(event.action)
+
+		// Can't do anything else if we didn't get a valid action object
+		if (action == null) {
+			return
+		}
+
+		// If this window isn't done yet add the action to the list
+		if (this.currentWindow.end == null) {
```

Comment:
Current window will be undefined when the end timestamp is set, so this is always true.

Reply:
removed